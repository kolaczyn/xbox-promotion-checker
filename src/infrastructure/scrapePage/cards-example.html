<li class="col mb-4 px-2">
  <div
    id="card-9-Gry na konsolę Xbox — oferty specjalnegame0"
    class="card h-100 material-card depth-4 depth-8-hover pb-4"
    data-bi-cn="Cyberpunk 2077"
    data-bi-hn="Gry na konsolę Xbox — oferty specjalne"
    data-bi-bhvr="0"
    data-bi-ct="Product Card"
    data-bi-pid="bx3m8l83bbrw"
    data-bi-prdname="Cyberpunk 2077"
    data-bi-carpos="9"
    data-bi-pa="Body"
    data-bi-compnm="Product Cards: Games"
    style="cursor: pointer"
  >
    <div class="img-flex-auto">
      <picture
        ><source
          srcset="
            https://store-images.s-microsoft.com/image/apps.47379.63407868131364914.bcaa868c-407e-42c2-baeb-48a3c9f29b54.89bb995b-b066-4a53-9fe4-0260ce07e894?q=90&amp;w=270&amp;h=405&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 540px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.47379.63407868131364914.bcaa868c-407e-42c2-baeb-48a3c9f29b54.89bb995b-b066-4a53-9fe4-0260ce07e894?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 860px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.47379.63407868131364914.bcaa868c-407e-42c2-baeb-48a3c9f29b54.89bb995b-b066-4a53-9fe4-0260ce07e894?q=90&amp;w=200&amp;h=300&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 1084px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.47379.63407868131364914.bcaa868c-407e-42c2-baeb-48a3c9f29b54.89bb995b-b066-4a53-9fe4-0260ce07e894?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 1400px)" />
        <img
          src="https://store-images.s-microsoft.com/image/apps.47379.63407868131364914.bcaa868c-407e-42c2-baeb-48a3c9f29b54.89bb995b-b066-4a53-9fe4-0260ce07e894?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7"
          class="card-img"
          alt=""
      /></picture>
    </div>
    <div
      class="card-header pt-3 px-3 d-flex flex-wrap"
      style="column-gap: 0.25rem"
    >
      <span class="badge bg-yellow mb-2 product-cards-savings-badge"
        >Oszczędzasz nawet 135,00&nbsp;zł</span
      ><span class="badge bg-black mb-2"></span>
    </div>
    <div class="card-body px-3">
      <h3 class="base mb-2 h6">
        <a
          href="https://www.microsoft.com/pl-pl/p/cyberpunk-2077/bx3m8l83bbrw"
          aria-label="Cyberpunk 2077"
          >Cyberpunk 2077</a
        >
      </h3>
      <p class="sr-only">Pierwotnie 269,99&nbsp;zł teraz 134,99&nbsp;zł</p>
      <p aria-hidden="true">
        <span class="text-line-through text-muted">269,99&nbsp;zł</span>
        <span class="font-weight-semibold">134,99&nbsp;zł</span>
      </p>
    </div>
  </div>
</li>

<li class="col mb-4 px-2">
  <div
    id="card-10-Gry na konsolę Xbox — oferty specjalnegame0"
    class="card h-100 material-card depth-4 depth-8-hover pb-4"
    data-bi-cn="Wiedźmin 3: Dziki Gon – Edycja Kompletna"
    data-bi-hn="Gry na konsolę Xbox — oferty specjalne"
    data-bi-bhvr="0"
    data-bi-ct="Product Card"
    data-bi-pid="c261457lcnmj"
    data-bi-prdname="Wiedźmin 3: Dziki Gon – Edycja Kompletna"
    data-bi-carpos="10"
    data-bi-pa="Body"
    data-bi-compnm="Product Cards: Games"
    style="cursor: pointer"
  >
    <div class="img-flex-auto">
      <picture
        ><source
          srcset="
            https://store-images.s-microsoft.com/image/apps.15193.65858607118306853.39ed2a08-df0d-4ae1-aee0-c66ffb783a34.7fe16441-021a-44c1-8af3-c7aa53940040?q=90&amp;w=270&amp;h=405&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 540px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.15193.65858607118306853.39ed2a08-df0d-4ae1-aee0-c66ffb783a34.7fe16441-021a-44c1-8af3-c7aa53940040?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 860px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.15193.65858607118306853.39ed2a08-df0d-4ae1-aee0-c66ffb783a34.7fe16441-021a-44c1-8af3-c7aa53940040?q=90&amp;w=200&amp;h=300&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 1084px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.15193.65858607118306853.39ed2a08-df0d-4ae1-aee0-c66ffb783a34.7fe16441-021a-44c1-8af3-c7aa53940040?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 1400px)" />
        <img
          src="https://store-images.s-microsoft.com/image/apps.15193.65858607118306853.39ed2a08-df0d-4ae1-aee0-c66ffb783a34.7fe16441-021a-44c1-8af3-c7aa53940040?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7"
          class="card-img"
          alt=""
      /></picture>
    </div>
    <div
      class="card-header pt-3 px-3 d-flex flex-wrap"
      style="column-gap: 0.25rem"
    >
      <span class="badge bg-yellow mb-2 product-cards-savings-badge"
        >Oszczędzasz nawet 100,00&nbsp;zł</span
      ><span class="badge bg-black mb-2"></span>
    </div>
    <div class="card-body px-3">
      <h3 class="base mb-2 h6">
        <a
          href="https://www.microsoft.com/pl-pl/p/wiedzmin-3-dziki-gon-edycja-kompletna/c261457lcnmj"
          aria-label="Wiedźmin 3: Dziki Gon – Edycja Kompletna"
          >Wiedźmin 3: Dziki Gon – Edycja Kompletna</a
        >
      </h3>
      <p class="sr-only">
        Pierwotnie 199,99&nbsp;zł teraz 99,99&nbsp;zł<span
          >Oferty zakupu w aplikacji</span
        >
      </p>
      <p aria-hidden="true">
        <span class="text-line-through text-muted">199,99&nbsp;zł</span>
        <span class="font-weight-semibold"
          >99,99&nbsp;zł<sup class="small"
            ><span aria-hidden="true">+</span></sup
          ></span
        >
      </p>
    </div>
  </div>
</li>

<li class="col mb-4 px-2">
  <div
    id="card-11-Gry na konsolę Xbox — oferty specjalnegame0"
    class="card h-100 material-card depth-4 depth-8-hover pb-4"
    data-bi-cn="Forza Horizon 5: Hot Wheels"
    data-bi-hn="Gry na konsolę Xbox — oferty specjalne"
    data-bi-bhvr="0"
    data-bi-ct="Product Card"
    data-bi-pid="9pnsz7gmwcqz"
    data-bi-prdname="Forza Horizon 5: Hot Wheels"
    data-bi-carpos="11"
    data-bi-pa="Body"
    data-bi-compnm="Product Cards: Games"
    style="cursor: pointer"
  >
    <div class="img-flex-auto">
      <picture
        ><source
          srcset="
            https://store-images.s-microsoft.com/image/apps.8266.14337700501983954.bd1bedb1-636e-4cb5-9fb9-6bda5bd998e8.cbfcac7c-2f17-4cdb-a1ca-e8e235900f30?q=90&amp;w=270&amp;h=405&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 540px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.8266.14337700501983954.bd1bedb1-636e-4cb5-9fb9-6bda5bd998e8.cbfcac7c-2f17-4cdb-a1ca-e8e235900f30?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 860px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.8266.14337700501983954.bd1bedb1-636e-4cb5-9fb9-6bda5bd998e8.cbfcac7c-2f17-4cdb-a1ca-e8e235900f30?q=90&amp;w=200&amp;h=300&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 1084px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.8266.14337700501983954.bd1bedb1-636e-4cb5-9fb9-6bda5bd998e8.cbfcac7c-2f17-4cdb-a1ca-e8e235900f30?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 1400px)" />
        <img
          src="https://store-images.s-microsoft.com/image/apps.8266.14337700501983954.bd1bedb1-636e-4cb5-9fb9-6bda5bd998e8.cbfcac7c-2f17-4cdb-a1ca-e8e235900f30?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7"
          class="card-img"
          alt=""
      /></picture>
    </div>
    <div
      class="card-header pt-3 px-3 d-flex flex-wrap"
      style="column-gap: 0.25rem"
    >
      <span class="badge bg-yellow mb-2 product-cards-savings-badge"
        >Oszczędzasz nawet 22,25&nbsp;zł</span
      ><span class="badge bg-black mb-2"></span>
    </div>
    <div class="card-body px-3">
      <h3 class="base mb-2 h6">
        <a
          href="https://www.microsoft.com/pl-pl/p/forza-horizon-5-hot-wheels/9pnsz7gmwcqz"
          aria-label="Forza Horizon 5: Hot Wheels"
          >Forza Horizon 5: Hot Wheels</a
        >
      </h3>
      <p class="sr-only">Pierwotnie 89,00&nbsp;zł teraz 66,75&nbsp;zł</p>
      <p aria-hidden="true">
        <span class="text-line-through text-muted">89,00&nbsp;zł</span>
        <span class="font-weight-semibold">66,75&nbsp;zł</span>
      </p>
    </div>
  </div>
</li>
