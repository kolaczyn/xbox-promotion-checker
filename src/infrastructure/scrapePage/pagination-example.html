<nav aria-label="Pagination for search results: 6 pages">
  <ul class="pagination">
    <li class="page-item disabled">
      <span class="page-link"><span>Previous</span></span>
    </li>
    <li class="page-item active">
      <a
        class="page-link"
        href="https://www.microsoft.com/pl-pl/store/deals/games/xbox"
        aria-current="page"
        ><span>1</span></a
      >
    </li>
    <li class="page-item">
      <a
        class="page-link"
        href="https://www.microsoft.com/pl-pl/store/deals/games/xbox?skipitems=90"
        ><span>2</span></a
      >
    </li>
    <li class="page-item">
      <a
        class="page-link"
        href="https://www.microsoft.com/pl-pl/store/deals/games/xbox?skipitems=180"
        ><span>3</span></a
      >
    </li>
    <li class="page-item">
      <a
        class="page-link"
        href="https://www.microsoft.com/pl-pl/store/deals/games/xbox?skipitems=270"
        ><span>4</span></a
      >
    </li>
    <li class="page-item">
      <a
        class="page-link"
        href="https://www.microsoft.com/pl-pl/store/deals/games/xbox?skipitems=360"
        ><span>5</span></a
      >
    </li>
    <li class="page-item">
      <a
        class="page-link"
        href="https://www.microsoft.com/pl-pl/store/deals/games/xbox?skipitems=450"
        ><span>6</span></a
      >
    </li>
    <li class="page-item">
      <a
        class="page-link"
        href="https://www.microsoft.com/pl-pl/store/deals/games/xbox?skipitems=90"
        ><span>Next</span></a
      >
    </li>
  </ul>
</nav>
