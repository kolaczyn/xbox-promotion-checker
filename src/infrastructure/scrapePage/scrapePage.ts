import parse, { HTMLElement } from 'node-html-parser'
import { GameInfo } from '../../types'
import { priceToNumber } from '../../utils/priceToNumber'

const extractNumberFromPrice = (str: string) =>
  str.replace(/&nbsp;|zÅ‚|\s|\+/g, '')

const scrapeCard = (card: HTMLElement): GameInfo | null => {
  const title = card.getAttribute('data-bi-cn')

  const prevEl = card.querySelector('.text-line-through.text-muted')
  const currEl = card.querySelector('.font-weight-semibold')

  const url = card.querySelector('a')?.getAttribute('href')
  const imageUrl = card.querySelector('img')?.getAttribute('src')

  const prevPriceStr = extractNumberFromPrice(prevEl?.innerText ?? '')
  const currPriceStr = extractNumberFromPrice(currEl?.innerText ?? '')

  const prevPrice = priceToNumber(prevPriceStr ?? '')
  const currPrice = priceToNumber(currPriceStr ?? '')

  if (!title || !prevPrice || !currPrice || !url || !imageUrl) {
    return null
  }

  return {
    title,
    prevPrice,
    currPrice,
    url,
    imageUrl,
  }
}

export const scrapePage = (html: string): GameInfo[] =>
  parse(html)
    .querySelectorAll('[data-bi-ct="Product Card"]')
    .map(scrapeCard)
    .filter((x): x is GameInfo => x !== null)

export const scrapeNumberOfPages = (html: string): number | null => {
  const paginationElements = parse(html).querySelectorAll('.page-item')
  return paginationElements.length > 0
    ? // `-2`, because of the 'next' and 'previous' links on the page
      paginationElements.length - 2
    : null
}
