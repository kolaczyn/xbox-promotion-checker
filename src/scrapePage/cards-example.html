<li class="col mb-4 px-2">
  <div
    class="card h-100 material-card depth-4 depth-8-hover pb-4"
    data-bi-cn="Rayman Legends"
    data-bi-hn="Gry na konsolę Xbox — oferty specjalne"
    data-bi-bhvr="0"
    data-bi-ct="Product Card"
    data-bi-pid="c26k4dvgr45b"
    data-bi-prdname="Rayman Legends"
    data-bi-carpos="0"
    data-bi-pa="Body"
    data-bi-compnm="Product Cards: Games"
    style="cursor: pointer"
    id="card-0-0"
  >
    <div class="img-flex-auto">
      <picture
        ><source
          srcset="
            https://store-images.s-microsoft.com/image/apps.25053.65843946604617087.4d3f1eaa-9f9b-4884-8da5-69d3fa77479f.c8e55bec-d1f6-4fe4-b3fa-13ac1e79814f?q=90&amp;w=270&amp;h=405&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 540px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.25053.65843946604617087.4d3f1eaa-9f9b-4884-8da5-69d3fa77479f.c8e55bec-d1f6-4fe4-b3fa-13ac1e79814f?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 860px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.25053.65843946604617087.4d3f1eaa-9f9b-4884-8da5-69d3fa77479f.c8e55bec-d1f6-4fe4-b3fa-13ac1e79814f?q=90&amp;w=200&amp;h=300&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 1084px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.25053.65843946604617087.4d3f1eaa-9f9b-4884-8da5-69d3fa77479f.c8e55bec-d1f6-4fe4-b3fa-13ac1e79814f?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7
          "
          media="(min-width: 1400px)" />
        <img
          src="https://store-images.s-microsoft.com/image/apps.25053.65843946604617087.4d3f1eaa-9f9b-4884-8da5-69d3fa77479f.c8e55bec-d1f6-4fe4-b3fa-13ac1e79814f?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%230078D7"
          class="card-img"
          alt=""
      /></picture>
    </div>
    <div
      class="card-header pt-3 px-3 d-flex flex-wrap"
      style="column-gap: 0.25rem"
    >
      <span class="badge bg-yellow mb-2 product-cards-savings-badge"
        >Oszczędzasz nawet 88,50 zł</span
      ><span class="badge bg-black mb-2"></span>
    </div>
    <div class="card-body px-3">
      <h2 class="base mb-2 h6">
        <a
          href="https://www.microsoft.com/pl-pl/p/rayman-legends/c26k4dvgr45b"
          aria-label="Rayman Legends"
          >Rayman Legends</a
        >
      </h2>
      <p class="sr-only">Pierwotnie 117,99 zł teraz 29,49 zł</p>
      <p aria-hidden="true">
        <span class="text-line-through text-muted">117,99 zł</span>
        <span class="font-weight-semibold">29,49 zł</span>
      </p>
    </div>
  </div>
</li>

<li class="col mb-4 px-2">
  <div
    class="card h-100 material-card depth-4 depth-8-hover pb-4"
    data-bi-cn="Assassin's Creed Valhalla Complete Edition"
    data-bi-hn="Gry na konsolę Xbox — oferty specjalne"
    data-bi-bhvr="0"
    data-bi-ct="Product Card"
    data-bi-pid="9np1b2v92qzx"
    data-bi-prdname="Assassin's Creed Valhalla Complete Edition"
    data-bi-carpos="1"
    data-bi-pa="Body"
    data-bi-compnm="Product Cards: Games"
    style="cursor: pointer"
    id="card-1-0"
  >
    <div class="img-flex-auto">
      <picture
        ><source
          srcset="
            https://store-images.s-microsoft.com/image/apps.61981.13959819632152004.cdf7fb2d-7428-47e4-82ed-ea0003da5b3a.bf2c464b-9236-4869-8423-14e7a6362433?q=90&amp;w=270&amp;h=405&amp;mode=crop&amp;format=jpg&amp;background=%23FFFFFF
          "
          media="(min-width: 540px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.61981.13959819632152004.cdf7fb2d-7428-47e4-82ed-ea0003da5b3a.bf2c464b-9236-4869-8423-14e7a6362433?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%23FFFFFF
          "
          media="(min-width: 860px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.61981.13959819632152004.cdf7fb2d-7428-47e4-82ed-ea0003da5b3a.bf2c464b-9236-4869-8423-14e7a6362433?q=90&amp;w=200&amp;h=300&amp;mode=crop&amp;format=jpg&amp;background=%23FFFFFF
          "
          media="(min-width: 1084px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.61981.13959819632152004.cdf7fb2d-7428-47e4-82ed-ea0003da5b3a.bf2c464b-9236-4869-8423-14e7a6362433?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%23FFFFFF
          "
          media="(min-width: 1400px)" />
        <img
          src="https://store-images.s-microsoft.com/image/apps.61981.13959819632152004.cdf7fb2d-7428-47e4-82ed-ea0003da5b3a.bf2c464b-9236-4869-8423-14e7a6362433?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%23FFFFFF"
          class="card-img"
          alt=""
      /></picture>
    </div>
    <div
      class="card-header pt-3 px-3 d-flex flex-wrap"
      style="column-gap: 0.25rem"
    >
      <span class="badge bg-yellow mb-2 product-cards-savings-badge"
        >Oszczędzasz nawet 386,75 zł</span
      ><span class="badge bg-black mb-2"></span>
    </div>
    <div class="card-body px-3">
      <h2 class="base mb-2 h6">
        <a
          href="https://www.microsoft.com/pl-pl/p/assassins-creed-valhalla-complete-edition/9np1b2v92qzx"
          aria-label="Assassin's Creed Valhalla Complete Edition"
          >Assassin's Creed Valhalla Complete Edition</a
        >
      </h2>
      <p class="sr-only">
        Pierwotnie 594,99 zł teraz 208,24 zł<span
          >Oferty zakupu w aplikacji</span
        >
      </p>
      <p aria-hidden="true">
        <span class="text-line-through text-muted">594,99 zł</span>
        <span class="font-weight-semibold"
          >208,24 zł<sup class="small"
            ><span aria-hidden="true">+</span></sup
          ></span
        >
      </p>
    </div>
  </div>
</li>
<li class="col mb-4 px-2">
  <div
    class="card h-100 material-card depth-4 depth-8-hover pb-4"
    data-bi-cn="Far Cry® 6 Gold Edition"
    data-bi-hn="Gry na konsolę Xbox — oferty specjalne"
    data-bi-bhvr="0"
    data-bi-ct="Product Card"
    data-bi-pid="9nt2tqpst9s0"
    data-bi-prdname="Far Cry® 6 Gold Edition"
    data-bi-carpos="2"
    data-bi-pa="Body"
    data-bi-compnm="Product Cards: Games"
    style="cursor: pointer"
    id="card-2-0"
  >
    <div class="img-flex-auto">
      <picture
        ><source
          srcset="
            https://store-images.s-microsoft.com/image/apps.33244.14422109142805611.71c71dc7-fe47-4083-ad56-ab54c9a5607d.32a28f70-8b6e-4b7e-ae64-c38ae01f32c1?q=90&amp;w=270&amp;h=405&amp;mode=crop&amp;format=jpg&amp;background=%23FFFFFF
          "
          media="(min-width: 540px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.33244.14422109142805611.71c71dc7-fe47-4083-ad56-ab54c9a5607d.32a28f70-8b6e-4b7e-ae64-c38ae01f32c1?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%23FFFFFF
          "
          media="(min-width: 860px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.33244.14422109142805611.71c71dc7-fe47-4083-ad56-ab54c9a5607d.32a28f70-8b6e-4b7e-ae64-c38ae01f32c1?q=90&amp;w=200&amp;h=300&amp;mode=crop&amp;format=jpg&amp;background=%23FFFFFF
          "
          media="(min-width: 1084px)" />
        <source
          srcset="
            https://store-images.s-microsoft.com/image/apps.33244.14422109142805611.71c71dc7-fe47-4083-ad56-ab54c9a5607d.32a28f70-8b6e-4b7e-ae64-c38ae01f32c1?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%23FFFFFF
          "
          media="(min-width: 1400px)" />
        <img
          src="https://store-images.s-microsoft.com/image/apps.33244.14422109142805611.71c71dc7-fe47-4083-ad56-ab54c9a5607d.32a28f70-8b6e-4b7e-ae64-c38ae01f32c1?q=90&amp;w=256&amp;h=384&amp;mode=crop&amp;format=jpg&amp;background=%23FFFFFF"
          class="card-img"
          alt=""
      /></picture>
    </div>
    <div
      class="card-header pt-3 px-3 d-flex flex-wrap"
      style="column-gap: 0.25rem"
    >
      <span class="badge bg-yellow mb-2 product-cards-savings-badge"
        >Oszczędzasz nawet 290,50 zł</span
      ><span class="badge bg-black mb-2"></span>
    </div>
    <div class="card-body px-3">
      <h2 class="base mb-2 h6">
        <a
          href="https://www.microsoft.com/pl-pl/p/far-cry-6-gold-edition/9nt2tqpst9s0"
          aria-label="Far Cry® 6 Gold Edition"
          >Far Cry® 6 Gold Edition</a
        >
      </h2>
      <p class="sr-only">
        Pierwotnie 414,99 zł teraz 124,49 zł<span
          >Oferty zakupu w aplikacji</span
        >
      </p>
      <p aria-hidden="true">
        <span class="text-line-through text-muted">414,99 zł</span>
        <span class="font-weight-semibold"
          >124,49 zł<sup class="small"
            ><span aria-hidden="true">+</span></sup
          ></span
        >
      </p>
    </div>
  </div>
</li>
