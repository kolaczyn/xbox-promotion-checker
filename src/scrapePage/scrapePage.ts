import parse, { HTMLElement } from 'node-html-parser'
import { GameInfo } from '../types'

const scrapeCard = (card: HTMLElement): GameInfo | null => {
  const title = card.getAttribute('data-bi-cn')
  const [prevEl, currEl] = card
    .querySelector('[aria-hidden="true"]')
    ?.querySelectorAll('span') ?? [null, null]
  const url = card.querySelector('a')?.getAttribute('href')
  const imageUrl = card.querySelector('img')?.getAttribute('src')

  const prevPrice = prevEl?.innerText
  const currPrice = currEl?.innerText.replace('+', '')

  if (!title || !prevPrice || !currPrice || !url || !imageUrl) {
    return null
  }

  return {
    title,
    prevPrice,
    currPrice,
    url,
    imageUrl,
  }
}

export const scrapePage = (html: string): GameInfo[] =>
  parse(html)
    .querySelectorAll('[data-bi-ct="Product Card"]')
    .map(scrapeCard)
    .filter((x): x is GameInfo => x !== null)

export const scrapeNumberOfPages = (html: string): number | null => {
  const paginationElements = parse(html).querySelectorAll('.page-item')
  return paginationElements.length > 0
    ? // `-2`, because of the 'next' and 'previous' links on the page
      paginationElements.length - 2
    : null
}
