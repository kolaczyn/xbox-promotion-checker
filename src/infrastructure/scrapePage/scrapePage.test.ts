import fs from 'fs/promises'
import { scrapeNumberOfPages, scrapePage } from './scrapePage'

const readHtml = async (htmlName: string) =>
  fs
    .readFile(`./src/infrastructure/scrapePage/${htmlName}`)
    .then(x => x.toString())

describe('scrapePage', () => {
  it('should retrieve correct game information', async () => {
    const html = await readHtml('cards-example.html')
    expect(scrapePage(html)).toMatchSnapshot()
  })
})

describe('scrapeNumberOfPages', () => {
  it('should get number of pages', async () => {
    const html = await readHtml('pagination-example.html')
    expect(scrapeNumberOfPages(html)).toEqual(6)
  })
  it('should return null if no element on the page', () => {
    const html = '<h1>hi mom</h1>'
    expect(scrapeNumberOfPages(html)).toBeNull()
  })
})
