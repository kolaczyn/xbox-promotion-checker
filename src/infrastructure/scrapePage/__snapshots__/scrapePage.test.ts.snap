// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrapePage should retrieve correct game information 1`] = `
[
  {
    "currPrice": 29.49,
    "imageUrl": "https://store-images.s-microsoft.com/image/apps.25053.65843946604617087.4d3f1eaa-9f9b-4884-8da5-69d3fa77479f.c8e55bec-d1f6-4fe4-b3fa-13ac1e79814f?q=90&w=256&h=384&mode=crop&format=jpg&background=%230078D7",
    "prevPrice": 117.99,
    "title": "Rayman Legends",
    "url": "https://www.microsoft.com/pl-pl/p/rayman-legends/c26k4dvgr45b",
  },
  {
    "currPrice": 208.24,
    "imageUrl": "https://store-images.s-microsoft.com/image/apps.61981.13959819632152004.cdf7fb2d-7428-47e4-82ed-ea0003da5b3a.bf2c464b-9236-4869-8423-14e7a6362433?q=90&w=256&h=384&mode=crop&format=jpg&background=%23FFFFFF",
    "prevPrice": 594.99,
    "title": "Assassin's Creed Valhalla Complete Edition",
    "url": "https://www.microsoft.com/pl-pl/p/assassins-creed-valhalla-complete-edition/9np1b2v92qzx",
  },
  {
    "currPrice": 124.49,
    "imageUrl": "https://store-images.s-microsoft.com/image/apps.33244.14422109142805611.71c71dc7-fe47-4083-ad56-ab54c9a5607d.32a28f70-8b6e-4b7e-ae64-c38ae01f32c1?q=90&w=256&h=384&mode=crop&format=jpg&background=%23FFFFFF",
    "prevPrice": 414.99,
    "title": "Far CryÂ® 6 Gold Edition",
    "url": "https://www.microsoft.com/pl-pl/p/far-cry-6-gold-edition/9nt2tqpst9s0",
  },
]
`;
